File: perf/benchmarks/micro/string_concat/string_concat.py
# String building — stdlib StringBuilder-style micro-benchmark.
#
# Build one large string by appending N small fragments using the language's
# recommended efficient builder (Python list + ''.join).
# Argument: N (default 500_000).

import sys
import time


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 500_000
    t0 = time.perf_counter_ns()
    parts: list[str] = []
    for i in range(n):
        parts.append("frag")
        parts.append(str(i))
        parts.append(";")
    s = "".join(parts)
    elapsed_ns = time.perf_counter_ns() - t0
    cs = len(s)
    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()

File: perf/benchmarks/micro/string_concat/string_concat.cpp
// String building — stdlib StringBuilder-style micro-benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <string>

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 500'000LL;

    auto t0 = std::chrono::steady_clock::now();
    std::string s;
    s.reserve(static_cast<size_t>(n) * 12);
    char buf[32];
    for (int64_t i = 0; i < n; ++i) {
        s.append("frag");
        int len = std::snprintf(buf, sizeof(buf), "%lld", static_cast<long long>(i));
        s.append(buf, static_cast<size_t>(len));
        s.push_back(';');
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
    std::printf("CHECKSUM:%zu\n", s.size());
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}
